### PR TITLE
Swift conversion

### PR DESCRIPTION
**Pull request by hhanesand**
ready for review :)

### CODE REVIEW

File: Pod/Classes/ABEGithubAPIClient.swift [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    
+    case MissingInformation = "Missing either the github token, repository name, or repository owner."
+    case JSONError = "There was an error converting the issue to JSON format."
+}
+
+final class ABEGithubAPIClient {
+    
+    public static var githubToken: String? = nil
+    public static var githubRepositoryName : String? = nil
+    public static var githubRepositoryOwner : String? = nil
```
Comment (heumn) [inline Pod/Classes/ABEGithubAPIClient.swift]: Nitty gritty :)

githubRepositoryOwner : 


githubRepositoryOwner:

File: Pod/Classes/ABEImageCollectionViewController.swift [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    static let kABEAddPictureCollectionViewCellIndex = 0
+    static let kABEAddPictureCollectionViewCellOffset = 1
+    
+    var issueManager: ABEIssueManager! {
+        didSet {
+//            issueManager.completionBlock = {
+//                Thread.abe_guaranteeBlockExecution { [weak self] in
+//                    self?.collectionView!.reloadData()
+//                }
+//            }
```
Comment (heumn) [inline Pod/Classes/ABEImageCollectionViewController.swift]: Commented code :)

File: Pod/Classes/ABEIssue.swift [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    var issueDescription: String?
+    
+    var imageURLS : [String] = []
+    
+    mutating func attachImage(withURL urlString: String) {
+        imageURLS.append(urlString)
+    }
+    
+    var textRepresentation : String? {
+        // TODO
```
Comment (heumn) [inline Pod/Classes/ABEIssue.swift]: Todo? What to do?

File: Pod/Classes/ABEReporterViewController.swift [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+        }
+    }
+    
+    public func saveIssue() {
+        
+        issueManager.issue.title = titleTextField.text
+        issueManager.issue.issueDescription = descriptionTextView.text
+
+        issueManager.saveIssue { [weak self] in
+            DispatchQueue.main.async { [weak self] in
```
Comment (heumn) [inline Pod/Classes/ABEReporterViewController.swift:92]: Is the duplicated [weak self] really needed here? 🤔

File: Pod/Classes/ABEissueManager.swift [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+            return uploadingImages.count != 0
+        }
+    }
+    
+    public private(set) var images: [UIImage] = []
+    public private(set) var localImageURLs: [URL] = []
+    
+    private var uploadingImages: [Data] = []
+    
+
```
Comment (heumn) [inline Pod/Classes/ABEissueManager.swift]: One line too many :)

File: Pod/Classes/ABEissueManager.swift [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    init(referenceView: UIView, delegate: ABEIssueManagerDelegate? = nil) {
+        self.referenceView = referenceView
+        self.delegate = delegate
+        
+        drawSnapshotOf(referenceView: referenceView) { [weak self] image in
+            self?.add(imageToIssue: image)
+        }
+    }
+    
+    private func drawSnapshotOf(referenceView view: UIView, complete: @escaping (UIImage) -> ()) {
```
Comment (heumn) [inline Pod/Classes/ABEissueManager.swift:60]: Please remind me that I have some code to draw the full content of the screen. Not just what is visible.... We should add that :)

File: Pod/Classes/ABEissueManager.swift [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+            }
+            
+            self?.uploadingImages.remove(at: index)
+            
+            if let `self` = self {
+                self.delegate?.issueManagerUploadingStateDidChange(issueManager: self)
+            }
+            
+            group.notify(queue: queue) {
+                // We are sure that the disk operation has finished
```
Comment (heumn) [inline Pod/Classes/ABEissueManager.swift]: 👍  👍

File: Pod/Classes/ABEissueManager.swift [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+                complete(url)
+            }
+        }
+    }
+    
+    public func saveIssue(completion: @escaping () -> ()) {
+        // self is only used in inner closure....
+        // TODO: Do we need [weak self] in both closures here? What happens if we remove the first [weak self]?
```
Comment (heumn) [inline Pod/Classes/ABEissueManager.swift]: No you are absolutely correct. [weak self] will not be required in the inner block. One thing I always think twice about when entering the scope of the inner block is, do I want to make sure I live throughout the scope of this? If so I often create a strong reference to self to make sure I get to finish my work. Just a thing to think of :)

If I wanted to retain self during the scope of the inner block

> guard let strongSelf = self else { return }

File: Pod/Classes/ABEImageCollectionViewController.swift (hunk 2 above)
Comment (hhanesand) [inline Pod/Classes/ABEImageCollectionViewController.swift]: It's removed in the other pull request 👍

File: Pod/Classes/ABEGithubAPIClient.swift (hunk 1 above)
Comment (hhanesand) [reply Pod/Classes/ABEGithubAPIClient.swift]: I changed this on the `error_handling` branch 👍

File: Pod/Classes/ABEImageCollectionViewController.swift (hunk 2 above)
Comment (hhanesand) [reply Pod/Classes/ABEImageCollectionViewController.swift]: This is fixed on the `error_handling` branch

File: Pod/Classes/ABEIssue.swift (hunk 3 above)
Comment (hhanesand) [reply Pod/Classes/ABEIssue.swift]: I was a todo for me to finish up `extraDebuggingInformation`, and I did it on the error_handling branch 👍

File: Pod/Classes/ABEissueManager.swift (hunk 5 above)
Comment (hhanesand) [reply Pod/Classes/ABEissueManager.swift]: Most of this has been removed on the `error_handling` branch

File: Pod/Classes/ABEissueManager.swift (hunk 6 above)
Comment (hhanesand) [reply Pod/Classes/ABEissueManager.swift:60]: yes will do

File: Pod/Classes/ABEissueManager.swift (hunk 8 above)
Comment (hhanesand) [reply Pod/Classes/ABEissueManager.swift]: good way of thinking of it!

`guard let`self`= self else { return }` also works if you want to keep using self

Comment (hhanesand) [reply Pod/Classes/ABEissueManager.swift]: This has been fixed on `error_handling` too :)

**Comment by hhanesand** [comment]
Ok, I've looked through all this and fixed it on the `error_handling` branch, which is a child branch of the `swift_conversion` branch, so when we merge `error_handling` in, the review comments will go away

**Comment by hhanesand** [comment]
It's 68982407978a57f5d9499f1d0bfe142dc313ad4d for reference

**Pull request merged**